### DISCUSSION TITLE
Suggestions for automated cert management?

### DISCUSSION THREAD

**Discussion by rustrust** [actix/actix-web#2995] [Q&A]
Does the actix-web community have any suggestions for how to perform automated cert management (e.g. ACME) with an actix web project?

For example, tide has this project:

https://github.com/http-rs/tide-acme

**Comment by robjtede** [comment]
We have an example showing this:
https://github.com/actix/examples/blob/master/https-tls/openssl-auto-le/src/main.rs

**Comment by Samielakkad** [comment]
For automated cert management, I would separate two cases:

1. Production behind a proxy/load balancer

The simplest and most common setup is to terminate TLS before Actix:

- Caddy with automatic HTTPS
- nginx + certbot
- Traefik
- cloud/load-balancer TLS such as AWS ALB, Cloudflare, Fly, etc.

Then Actix listens on plain HTTP internally. This is usually the least fragile production setup.

2. Actix process manages certificates itself

If the Actix process must serve TLS directly, use an ACME client/library to provision certs and configure Rustls/OpenSSL with the generated files. The Actix examples repo has an OpenSSL Let's Encrypt style example, but I would still prefer an external TLS terminator unless you specifically need in-process cert management.

Operationally, cert automation needs more than just "get cert":

- HTTP-01 or DNS-01 challenge handling
- persistent certificate storage
- renewal before expiry
- reloading TLS config without downtime, or restarting safely
- rate-limit handling

That is why Caddy/Traefik/certbot are attractive: they already solved those edges.

So my recommendation: use Caddy/Traefik/nginx-certbot unless you have a strong reason to keep ACME inside the Rust binary.
